File: 22_专题/离线查询/根号分治/RightMostLeftMostQuery-线段树.ts
/* eslint-disable no-inner-declarations */

// 对每个下标，查询 最右侧/最左侧/右侧第一个/左侧第一个 lower/floor/ceiling/higher 的元素.
// 动态单调栈(DynamicMonoStack).
// 线段树实现(非常快).

import { SegmentTreeRangeAddRangeMinMax } from '../../../6_tree/线段树/template/atcoder_segtree/hot/SegmentTreeRangeAddRangeMinMax-区间加区间最大最小值'

class RightMostLeftMostQuerySegmentTree {
  private readonly _n: number
  private readonly _rangeAddRangeMinMax: SegmentTreeRangeAddRangeMinMax

  constructor(arr: ArrayLike<number>) {
    this._n = arr.length
    this._rangeAddRangeMinMax = new SegmentTreeRangeAddRangeMinMax(arr)
  }

  set(index: number, value: number): void {
    if (index < 0 || index >= this._n) return
    this._rangeAddRangeMinMax.set(index, value)
  }

  addRange(start: number, end: number, delta: number): void {
    if (start < 0) start = 0
    if (end > this._n) end = this._n
    if (start >= end) return
    this._rangeAddRangeMinMax.update(start, end, delta)
  }

  /**
   * 查询`index`右侧最远的下标`j`，使得 `nums[j] < nums[index]`.
   * 如果不存在，返回`-1`.
   */
  rightMostLower(index: number): number {
    const cur = this._rangeAddRangeMinMax.get(index)
    const cand = this._rangeAddRangeMinMax.minLeft(this._n, min => min >= cur) - 1
    return cand > index ? cand : -1
  }

  /**
   * 查询`index`右侧最远的下标`j`，使得 `nums[j] <= nums[index]`.
   * 如果不存在，返回`-1`.
   */
  rightMostFloor(index: number): number {
    const cur = this._rangeAddRangeMinMax.get(index)
    const cand = this._rangeAddRangeMinMax.minLeft(this._n, min => min > cur) - 1
    return cand > index ? cand : -1
  }

  /**
   * 查询`index`右侧最远的下标`j`，使得 `nums[j] >= nums[index]`.
   * 如果不存在，返回`-1`.
   */
  rightMostCeiling(index: number): number {
    const cur = this._rangeAddRangeMinMax.get(index)
    const cand = this._rangeAddRangeMinMax.minLeft(this._n, (_, max) => max < cur) - 1
    return cand > index ? cand : -1
  }

  /**
   * 查询`index`右侧最远的下标`j`，使得 `nums[j] > nums[index]`.
   * 如果不存在，返回`-1`.
   */
  rightMostHigher(index: number): number {
    const cur = this._rangeAddRangeMinMax.get(index)
    const cand = this._rangeAddRangeMinMax.minLeft(this._n, (_, max) => max <= cur) - 1
    return cand > index ? cand : -1
  }

  /**
   * 查询`index`左侧最远的下标`j`，使得 `nums[j] < nums[index]`.
   * 如果不存在，返回`-1`.
   */
  leftMostLower(index: number): number {
    const cur = this._rangeAddRangeMinMax.get(index)
    const cand = this._rangeAddRangeMinMax.maxRight(0, min => min >= cur)
    return cand < index ? cand : -1
  }

  /**
   * 查询`index`左侧最远的下标`j`，使得 `nums[j] <= nums[index]`.
   * 如果不存在，返回`-1`.
   */
  leftMostFloor(index: number): number {
    const cur = this._rangeAddRangeMinMax.get(index)
    const cand = this._rangeAddRangeMinMax.maxRight(0, min => min > cur)
    return cand < index ? cand : -1
  }

  /**
   * 查询`index`左侧最远的下标`j`，使得 `nums[j] >= nums[index]`.
   * 如果不存在，返回`-1`.
   */
  leftMostCeiling(index: number): number {
    const cur = this._rangeAddRangeMinMax.get(index)
    const cand = this._rangeAddRangeMinMax.maxRight(0, (_, max) => max < cur)
    return cand < index ? cand : -1
  }

  /**
   * 查询`index`左侧最远的下标`j`，使得 `nums[j] > nums[index]`.
   * 如果不存在，返回`-1`.
   */
  leftMostHigher(index: number): number {
    const cur = this._rangeAddRangeMinMax.get(index)
    const cand = this._rangeAddRangeMinMax.maxRight(0, (_, max) => max <= cur)
    return cand < index ? cand : -1
  }

  /**
   * 查询`index`右侧最近的下标`j`，使得 `nums[j] < nums[index]`.
   * 如果不存在，返回`-1`.
   */
  rightNearestLower(index: number): number {
    const cur = this._rangeAddRangeMinMax.get(index)
    const cand = this._rangeAddRangeMinMax.maxRight(index + 1, min => min >= cur)
    return cand === this._n ? -1 : cand
  }

  /**
   * 查询`index`右侧最近的下标`j`，使得 `nums[j] <= nums[index]`.
   * 如果不存在，返回`-1`.
   */
  rightNearestFloor(index: number): number {
    const cur = this._rangeAddRangeMinMax.get(index)
    const cand = this._rangeAddRangeMinMax.maxRight(index + 1, min => min > cur)
    return cand === this._n ? -1 : cand
  }

  /**
   * 查询`index`右侧最近的下标`j`，使得 `nums[j] >= nums[index]`.
   * 如果不存在，返回`-1`.
   */
  rightNearestCeiling(index: number): number {
    const cur = this._rangeAddRangeMinMax.get(index)
    const cand = this._rangeAddRangeMinMax.maxRight(index + 1, (_, max) => max < cur)
    return cand === this._n ? -1 : cand
  }

  /**
   * 查询`index`右侧最近的下标`j`，使得 `nums[j] > nums[index]`.
   * 如果不存在，返回`-1`.
   */
  rightNearestHigher(index: number): number {
    const cur = this._rangeAddRangeMinMax.get(index)
    const cand = this._rangeAddRangeMinMax.maxRight(index + 1, (_, max) => max <= cur)
    return cand === this._n ? -1 : cand
  }

  /**
   * 查询`index`左侧最近的下标`j`，使得 `nums[j] < nums[index]`.
   * 如果不存在，返回`-1`.
   */
  leftNearestLower(index: number): number {
    const cur = this._rangeAddRangeMinMax.get(index)
    const cand = this._rangeAddRangeMinMax.minLeft(index, min => min >= cur) - 1
    return cand
  }

  /**
   * 查询`index`左侧最近的下标`j`，使得 `nums[j] <= nums[index]`.
   * 如果不存在，返回`-1`.
   */
  leftNearestFloor(index: number): number {
    const cur = this._rangeAddRangeMinMax.get(index)
    const cand = this._rangeAddRangeMinMax.minLeft(index, min => min > cur) - 1
    return cand
  }

  /**
   * 查询`index`左侧最近的下标`j`，使得 `nums[j] >= nums[index]`.
   * 如果不存在，返回`-1`.
   */
  leftNearestCeiling(index: number): number {
    const cur = this._rangeAddRangeMinMax.get(index)
    const cand = this._rangeAddRangeMinMax.minLeft(index, (_, max) => max < cur) - 1
    return cand
  }

  /**
   * 查询`index`左侧最近的下标`j`，使得 `nums[j] > nums[index]`.
   * 如果不存在，返回`-1`.
   */
  leftNearestHigher(index: number): number {
    const cur = this._rangeAddRangeMinMax.get(index)
    const cand = this._rangeAddRangeMinMax.minLeft(index, (_, max) => max <= cur) - 1
    return cand
  }
}

if (require.main === module) {
  // 962. 最大宽度坡
  // https://leetcode.cn/problems/maximum-width-ramp/
  function maxWidthRamp(nums: number[]): number {
    const Q = new RightMostLeftMostQuerySegmentTree(nums)
    let res = 0
    for (let i = 0; i < nums.length; i++) {
      const cand = Q.rightMostCeiling(i)
      if (cand !== -1) {
        res = Math.max(res, cand - i)
      }
    }
    return res
  }

  // 2863. 最长半递减数组
  // https://leetcode.cn/problems/maximum-length-of-semi-decreasing-subarrays/description/
  function maxSubarrayLength(nums: number[]): number {
    const Q = new RightMostLeftMostQuerySegmentTree(nums)
    let res = 0
    for (let i = 0; i < nums.length; i++) {
      const cand = Q.rightMostLower(i)
      if (cand !== -1) {
        res = Math.max(res, cand - i + 1)
      }
    }
    return res
  }

  // 901. 股票价格跨度
  // 当日股票价格的 跨度 被定义为股票价格小于或等于今天价格的最大连续日数
  class StockSpanner {
    private readonly _Q: RightMostLeftMostQuerySegmentTree
    private _ptr = 0

    constructor() {
      this._Q = new RightMostLeftMostQuerySegmentTree(Array(1e5 + 10).fill(0))
    }

    next(price: number): number {
      const pos = this._ptr++
      this._Q.set(pos, price)
      const higherPos = this._Q.leftNearestHigher(pos)
      return higherPos === -1 ? pos + 1 : pos - higherPos
    }
  }

  // checkWithBruteForce()
  function checkWithBruteForce(): void {
    class Mocker {
      readonly _nums: number[]
      constructor(nums: number[]) {
        this._nums = nums
      }

      set(index: number, value: number): void {
        this._nums[index] = value
      }

      addRange(start: number, end: number, delta: number): void {
        for (let i = start; i < end; i++) this._nums[i] += delta
      }

      rightMostLower(index: number): number {
        for (let i = this._nums.length - 1; i > index; i--) {
          if (this._nums[i] < this._nums[index]) return i
        }
        return -1
      }

      rightMostFloor(index: number): number {
        for (let i = this._nums.length - 1; i > index; i--) {
          if (this._nums[i] <= this._nums[index]) return i
        }
        return -1
      }

      rightMostCeiling(index: number): number {
        for (let i = this._nums.length - 1; i > index; i--) {
          if (this._nums[i] >= this._nums[index]) return i
        }
        return -1
      }

      rightMostHigher(index: number): number {
        for (let i = this._nums.length - 1; i > index; i--) {
          if (this._nums[i] > this._nums[index]) return i
        }
        return -1
      }

      leftMostLower(index: number): number {
        for (let i = 0; i < index; i++) {
          if (this._nums[i] < this._nums[index]) return i
        }
        return -1
      }

      leftMostFloor(index: number): number {
        for (let i = 0; i < index; i++) {
          if (this._nums[i] <= this._nums[index]) return i
        }
        return -1
      }

      leftMostCeiling(index: number): number {
        for (let i = 0; i < index; i++) {
          if (this._nums[i] >= this._nums[index]) return i
        }
        return -1
      }

      leftMostHigher(index: number): number {
        for (let i = 0; i < index; i++) {
          if (this._nums[i] > this._nums[index]) return i
        }
        return -1
      }

      rightNearestLower(index: number): number {
        for (let i = index + 1; i < this._nums.length; i++) {
          if (this._nums[i] < this._nums[index]) return i
        }
        return -1
      }

      rightNearestFloor(index: number): number {
        for (let i = index + 1; i < this._nums.length; i++) {
          if (this._nums[i] <= this._nums[index]) return i
        }
        return -1
      }

      rightNearestCeiling(index: number): number {
        for (let i = index + 1; i < this._nums.length; i++) {
          if (this._nums[i] >= this._nums[index]) return i
        }
        return -1
      }

      rightNearestHigher(index: number): number {
        for (let i = index + 1; i < this._nums.length; i++) {
          if (this._nums[i] > this._nums[index]) return i
        }
        return -1
      }

      leftNearestLower(index: number): number {
        for (let i = index - 1; i >= 0; i--) {
          if (this._nums[i] < this._nums[index]) return i
        }
        return -1
      }

      leftNearestFloor(index: number): number {
        for (let i = index - 1; i >= 0; i--) {
          if (this._nums[i] <= this._nums[index]) return i
        }
        return -1
      }

      leftNearestCeiling(index: number): number {
        for (let i = index - 1; i >= 0; i--) {
          if (this._nums[i] >= this._nums[index]) return i
        }
        return -1
      }

      leftNearestHigher(index: number): number {
        for (let i = index - 1; i >= 0; i--) {
          if (this._nums[i] > this._nums[index]) return i
        }
        return -1
      }
    }

    const N = 1e5
    const MAX = 1e9
    const randomNums = Array.from({ length: N }, () => (Math.random() * MAX) | 0)
    const mocker = new Mocker(randomNums.slice())
    const real = new RightMostLeftMostQuerySegmentTree(randomNums)
    const debug = (
      mockerFunc: (index: number) => number,
      realFunc: (index: number) => number
    ): void => {
      const index = randint(0, randomNums.length - 1)
      const mockerRes = mockerFunc(index)
      const QRes = realFunc(index)
      if (mockerRes !== QRes) {
        console.log(realFunc.name, index, mockerRes, QRes)
        // console.log(mocker._nums, real._nums, real._blockLazy)
        // console.log('blockMin', real._blockMin)
        // console.log('blockMax', real._blockMax)
        throw new Error(realFunc.name)
      }
    }

    const randint = (a: number, b: number) => Math.floor(Math.random() * (b - a + 1)) + a
    for (let i = 0; i < 1e5; i++) {
      const op = randint(0, 17)
      // set
      if (op === 0) {
        const index = randint(0, randomNums.length - 1)
        const value = (Math.random() * MAX) | 0
        mocker.set(index, value)
        real.set(index, value)
      } else if (op === 1) {
        // addRange
        const start = randint(0, randomNums.length - 1)
        const end = randint(start, randomNums.length - 1)
        const delta = (Math.random() * MAX) | 0
        mocker.addRange(start, end, delta)
        real.addRange(start, end, delta)
      } else if (op === 2) {
        debug(mocker.rightMostLower.bind(mocker), real.rightMostLower.bind(real))
      } else if (op === 3) {
        debug(mocker.rightMostFloor.bind(mocker), real.rightMostFloor.bind(real))
      } else if (op === 4) {
        debug(mocker.rightMostCeiling.bind(mocker), real.rightMostCeiling.bind(real))
      } else if (op === 5) {
        debug(mocker.rightMostHigher.bind(mocker), real.rightMostHigher.bind(real))
      } else if (op === 6) {
        debug(mocker.leftMostLower.bind(mocker), real.leftMostLower.bind(real))
      } else if (op === 7) {
        debug(mocker.leftMostFloor.bind(mocker), real.leftMostFloor.bind(real))
      } else if (op === 8) {
        debug(mocker.leftMostCeiling.bind(mocker), real.leftMostCeiling.bind(real))
      } else if (op === 9) {
        debug(mocker.leftMostHigher.bind(mocker), real.leftMostHigher.bind(real))
      } else if (op === 10) {
        debug(mocker.rightNearestLower.bind(mocker), real.rightNearestLower.bind(real))
      } else if (op === 11) {
        debug(mocker.rightNearestFloor.bind(mocker), real.rightNearestFloor.bind(real))
      } else if (op === 12) {
        debug(mocker.rightNearestCeiling.bind(mocker), real.rightNearestCeiling.bind(real))
      } else if (op === 13) {
        debug(mocker.rightNearestHigher.bind(mocker), real.rightNearestHigher.bind(real))
      } else if (op === 14) {
        debug(mocker.leftNearestLower.bind(mocker), real.leftNearestLower.bind(real))
      } else if (op === 15) {
        debug(mocker.leftNearestFloor.bind(mocker), real.leftNearestFloor.bind(real))
      } else if (op === 16) {
        debug(mocker.leftNearestCeiling.bind(mocker), real.leftNearestCeiling.bind(real))
      } else if (op === 17) {
        debug(mocker.leftNearestHigher.bind(mocker), real.leftNearestHigher.bind(real))
      }
    }

    console.log('ok')
  }

  testTime()
  function testTime(): void {
    const bigArr = Array(1e5)
      .fill(0)
      .map(() => (Math.random() * 1e9) | 0)
    const Q = new RightMostLeftMostQuerySegmentTree(bigArr)
    console.time('bigArr')
    for (let i = 0; i < 1e5; i++) {
      Q.set(i, i)
      Q.addRange(i, i + 1, i)
      Q.rightMostLower(i)
      Q.rightMostFloor(i)
      Q.rightMostCeiling(i)
      Q.rightMostHigher(i)
      Q.leftMostLower(i)
      Q.leftMostFloor(i)
      Q.leftMostCeiling(i)
      Q.leftMostHigher(i)
      Q.rightNearestLower(i)
      Q.rightNearestFloor(i)
      Q.rightNearestCeiling(i)
      Q.rightNearestHigher(i)
      Q.leftNearestLower(i)
      Q.leftNearestFloor(i)
      Q.leftNearestCeiling(i)
      Q.leftNearestHigher(i)
    }
    console.timeEnd('bigArr') // bigArr: 330.583ms (比分块的2s快)
  }
}

File: 22_专题/离线查询/根号分治/RightMostLeftMostQuery-线段树.go
/* eslint-disable no-inner-declarations */

// !推荐使用，更快.
// 对每个下标，查询 最右侧/最左侧/右侧第一个/左侧第一个 lower/floor/ceiling/higher 的元素.
// 动态单调栈(DynamicMonoStack、MonoStackDynamic).
// 线段树实现.
// !左侧右侧包含当前位置.

package main

import (
	"bufio"
	"fmt"
	"math/bits"
	"os"
	"strings"
)

func main() {
	abc382c()
}

// C - Kaiten Sushi
// https://atcoder.jp/contests/abc382/tasks/abc382_c
// 回转寿司
//
// 在某个回转寿司店，有 N 人从 1 到 N 的编号到访。人 i 的美食度是 A i ​ 。
// 传送带上流动着 M 个寿司。第 j 个流动的寿司的美味程度是 B j ​ 。
// 每个寿司依次流过每个人 1,2,…,N 的面前。
// 每个人在美味程度>=自己美食度的寿司流到自己面前时会取走并食用该寿司，其他情况下则不做任何事情。
// 请确定每个寿司由谁食用，或者是否没有人食用。
//
// 等价于：对于每个寿司，找到第一个美食度大于等于它的人。
func abc382c() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var N, M int32
	fmt.Fscan(in, &N, &M)
	A, B := make([]int, N), make([]int, M)
	for i := int32(0); i < N; i++ {
		fmt.Fscan(in, &A[i])
	}
	for i := int32(0); i < M; i++ {
		fmt.Fscan(in, &B[i])
	}

	res := make([]int32, M)
	Q := NewRightMostLeftMostQuery(A)
	for i := int32(0); i < M; i++ {
		res[i] = int32(Q.RightNearestFloor(0, B[i]))
	}
	for _, v := range res {
		if v == -1 {
			fmt.Fprintln(out, -1)
		} else {
			fmt.Fprintln(out, v+1)
		}
	}
}

const INF int = 1e18

type RightMostLeftMostQuery struct {
	_n    int
	_tree *SegmentTreeRangeAddRangeMinMax
}

func NewRightMostLeftMostQuery(arr []int) *RightMostLeftMostQuery {
	n := len(arr)
	tree := NewSegmentTreeRangeAddRangeFrom(n, func(i int) E { return E{min: arr[i], max: arr[i]} })
	return &RightMostLeftMostQuery{_n: n, _tree: tree}
}

func (rm *RightMostLeftMostQuery) Get(index int) int {
	if index < 0 || index >= rm._n {
		panic(fmt.Sprintf("index %v out of range [0, %v)", index, rm._n))
	}
	return rm._tree.Get(index).min
}

func (rm *RightMostLeftMostQuery) Set(index int, value int) {
	if 0 <= index && index < rm._n {
		rm._tree.Set(index, E{min: value, max: value})
	}
}

func (rm *RightMostLeftMostQuery) AddRange(start, end int, delta int) {
	if start < 0 {
		start = 0
	}
	if end > rm._n {
		end = rm._n
	}
	if start >= end {
		return
	}
	rm._tree.Update(start, end, delta)
}

func (rm *RightMostLeftMostQuery) RightMostLower(index int, target int) int {
	cand := rm._tree.MinLeft(rm._n, func(e E) bool { return e.min >= target }) - 1
	if cand >= index {
		return cand
	}
	return -1
}

func (rm *RightMostLeftMostQuery) RightMostFloor(index int, target int) int {
	cand := rm._tree.MinLeft(rm._n, func(e E) bool { return e.min > target }) - 1
	if cand >= index {
		return cand
	}
	return -1
}

func (rm *RightMostLeftMostQuery) RightMostCeiling(index int, target int) int {
	cand := rm._tree.MinLeft(rm._n, func(e E) bool { return e.max < target }) - 1
	if cand >= index {
		return cand
	}
	return -1
}

func (rm *RightMostLeftMostQuery) RightMostHigher(index int, target int) int {
	cand := rm._tree.MinLeft(rm._n, func(e E) bool { return e.max <= target }) - 1
	if cand >= index {
		return cand
	}
	return -1
}

func (rm *RightMostLeftMostQuery) LeftMostLower(index int, target int) int {
	cand := rm._tree.MaxRight(0, func(e E) bool { return e.min >= target })
	if cand <= index {
		return cand
	}
	return -1
}

func (rm *RightMostLeftMostQuery) LeftMostFloor(index int, target int) int {
	cand := rm._tree.MaxRight(0, func(e E) bool { return e.min > target })
	if cand <= index {
		return cand
	}
	return -1
}

func (rm *RightMostLeftMostQuery) LeftMostCeiling(index int, target int) int {
	cand := rm._tree.MaxRight(0, func(e E) bool { return e.max < target })
	if cand <= index {
		return cand
	}
	return -1
}

func (rm *RightMostLeftMostQuery) LeftMostHigher(index int, target int) int {
	cand := rm._tree.MaxRight(0, func(e E) bool { return e.max <= target })
	if cand <= index {
		return cand
	}
	return -1
}

func (rm *RightMostLeftMostQuery) RightNearestLower(index int, target int) int {
	cand := rm._tree.MaxRight(index, func(e E) bool { return e.min >= target })
	if cand < rm._n {
		return cand
	}
	return -1
}

func (rm *RightMostLeftMostQuery) RightNearestFloor(index int, target int) int {
	cand := rm._tree.MaxRight(index, func(e E) bool { return e.min > target })
	if cand < rm._n {
		return cand
	}
	return -1
}

func (rm *RightMostLeftMostQuery) RightNearestCeiling(index int, target int) int {
	cand := rm._tree.MaxRight(index, func(e E) bool { return e.max < target })
	if cand < rm._n {
		return cand
	}
	return -1
}

func (rm *RightMostLeftMostQuery) RightNearestHigher(index int, target int) int {
	cand := rm._tree.MaxRight(index, func(e E) bool { return e.max <= target })
	if cand < rm._n {
		return cand
	}
	return -1
}

func (rm *RightMostLeftMostQuery) LeftNearestLower(index int, target int) int {
	cand := rm._tree.MinLeft(index+1, func(e E) bool { return e.min >= target }) - 1
	return cand
}

func (rm *RightMostLeftMostQuery) LeftNearestFloor(index int, target int) int {
	cand := rm._tree.MinLeft(index+1, func(e E) bool { return e.min > target }) - 1
	return cand
}

func (rm *RightMostLeftMostQuery) LeftNearestCeiling(index int, target int) int {
	cand := rm._tree.MinLeft(index+1, func(e E) bool { return e.max < target }) - 1
	return cand
}

func (rm *RightMostLeftMostQuery) LeftNearestHigher(index int, target int) int {
	cand := rm._tree.MinLeft(index+1, func(e E) bool { return e.max <= target }) - 1
	return cand
}

// RangeAddRangeMinMax-区间加区间最大最小值
type E = struct{ min, max int }
type Id = int

func (*SegmentTreeRangeAddRangeMinMax) e() E   { return E{min: INF, max: -INF} }
func (*SegmentTreeRangeAddRangeMinMax) id() Id { return 0 }
func (*SegmentTreeRangeAddRangeMinMax) op(left, right E) E {
	left.min = min(left.min, right.min)
	left.max = max(left.max, right.max)
	return left
}
func (*SegmentTreeRangeAddRangeMinMax) mapping(f Id, g E) E {
	if f == 0 {
		return g
	}
	g.min += f
	g.max += f
	return g
}
func (*SegmentTreeRangeAddRangeMinMax) composition(f, g Id) Id {
	return f + g
}
func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
func max(a, b int) int {
	if a < b {
		return b
	}
	return a
}

// !template
type SegmentTreeRangeAddRangeMinMax struct {
	n    int
	size int
	log  int
	data []E
	lazy []Id
}

func NewSegmentTreeRangeAddRangeFrom(n int, f func(i int) E) *SegmentTreeRangeAddRangeMinMax {
	tree := &SegmentTreeRangeAddRangeMinMax{}
	tree.n = n
	tree.log = int(bits.Len(uint(n - 1)))
	tree.size = 1 << tree.log
	tree.data = make([]E, tree.size<<1)
	tree.lazy = make([]Id, tree.size)
	for i := range tree.data {
		tree.data[i] = tree.e()
	}
	for i := range tree.lazy {
		tree.lazy[i] = tree.id()
	}
	for i := 0; i < n; i++ {
		tree.data[tree.size+i] = f(i)
	}
	for i := tree.size - 1; i >= 1; i-- {
		tree.pushUp(i)
	}
	return tree
}

// 查询切片[left:right]的值
//
//	0<=left<=right<=len(tree.data)
func (tree *SegmentTreeRangeAddRangeMinMax) Query(left, right int) E {
	if left < 0 {
		left = 0
	}
	if right > tree.n {
		right = tree.n
	}
	if left >= right {
		return tree.e()
	}
	left += tree.size
	right += tree.size
	for i := tree.log; i >= 1; i-- {
		if ((left >> i) << i) != left {
			tree.pushDown(left >> i)
		}
		if ((right >> i) << i) != right {
			tree.pushDown((right - 1) >> i)
		}
	}
	sml, smr := tree.e(), tree.e()
	for left < right {
		if left&1 != 0 {
			sml = tree.op(sml, tree.data[left])
			left++
		}
		if right&1 != 0 {
			right--
			smr = tree.op(tree.data[right], smr)
		}
		left >>= 1
		right >>= 1
	}
	return tree.op(sml, smr)
}
func (tree *SegmentTreeRangeAddRangeMinMax) QueryAll() E {
	return tree.data[1]
}

// 更新切片[left:right]的值
//
//	0<=left<=right<=len(tree.data)
func (tree *SegmentTreeRangeAddRangeMinMax) Update(left, right int, f Id) {
	if left < 0 {
		left = 0
	}
	if right > tree.n {
		right = tree.n
	}
	if left >= right {
		return
	}
	left += tree.size
	right += tree.size
	for i := tree.log; i >= 1; i-- {
		if ((left >> i) << i) != left {
			tree.pushDown(left >> i)
		}
		if ((right >> i) << i) != right {
			tree.pushDown((right - 1) >> i)
		}
	}
	l2, r2 := left, right
	for left < right {
		if left&1 != 0 {
			tree.propagate(left, f)
			left++
		}
		if right&1 != 0 {
			right--
			tree.propagate(right, f)
		}
		left >>= 1
		right >>= 1
	}
	left = l2
	right = r2
	for i := 1; i <= tree.log; i++ {
		if ((left >> i) << i) != left {
			tree.pushUp(left >> i)
		}
		if ((right >> i) << i) != right {
			tree.pushUp((right - 1) >> i)
		}
	}
}

// 二分查询最小的 left 使得切片 [left:right] 内的值满足 predicate
func (tree *SegmentTreeRangeAddRangeMinMax) MinLeft(right int, predicate func(data E) bool) int {
	if right == 0 {
		return 0
	}
	right += tree.size
	for i := tree.log; i >= 1; i-- {
		tree.pushDown((right - 1) >> i)
	}
	res := tree.e()
	for {
		right--
		for right > 1 && right&1 != 0 {
			right >>= 1
		}
		if !predicate(tree.op(tree.data[right], res)) {
			for right < tree.size {
				tree.pushDown(right)
				right = right<<1 | 1
				if tmp := tree.op(tree.data[right], res); predicate(tmp) {
					res = tmp
					right--
				}
			}
			return right + 1 - tree.size
		}
		res = tree.op(tree.data[right], res)
		if (right & -right) == right {
			break
		}
	}
	return 0
}

// 二分查询最大的 right 使得切片 [left:right] 内的值满足 predicate
func (tree *SegmentTreeRangeAddRangeMinMax) MaxRight(left int, predicate func(data E) bool) int {
	if left == tree.n {
		return tree.n
	}
	left += tree.size
	for i := tree.log; i >= 1; i-- {
		tree.pushDown(left >> i)
	}
	res := tree.e()
	for {
		for left&1 == 0 {
			left >>= 1
		}
		if !predicate(tree.op(res, tree.data[left])) {
			for left < tree.size {
				tree.pushDown(left)
				left <<= 1
				if tmp := tree.op(res, tree.data[left]); predicate(tmp) {
					res = tmp
					left++
				}
			}
			return left - tree.size
		}
		res = tree.op(res, tree.data[left])
		left++
		if (left & -left) == left {
			break
		}
	}
	return tree.n
}

// 单点查询(不需要 pushUp/op 操作时使用)
func (tree *SegmentTreeRangeAddRangeMinMax) Get(index int) E {
	index += tree.size
	for i := tree.log; i >= 1; i-- {
		tree.pushDown(index >> i)
	}
	return tree.data[index]
}

// 单点赋值
func (tree *SegmentTreeRangeAddRangeMinMax) Set(index int, e E) {
	index += tree.size
	for i := tree.log; i >= 1; i-- {
		tree.pushDown(index >> i)
	}
	tree.data[index] = e
	for i := 1; i <= tree.log; i++ {
		tree.pushUp(index >> i)
	}
}

func (tree *SegmentTreeRangeAddRangeMinMax) pushUp(root int) {
	tree.data[root] = tree.op(tree.data[root<<1], tree.data[root<<1|1])
}
func (tree *SegmentTreeRangeAddRangeMinMax) pushDown(root int) {
	if tree.lazy[root] != tree.id() {
		tree.propagate(root<<1, tree.lazy[root])
		tree.propagate(root<<1|1, tree.lazy[root])
		tree.lazy[root] = tree.id()
	}
}
func (tree *SegmentTreeRangeAddRangeMinMax) propagate(root int, f Id) {
	tree.data[root] = tree.mapping(f, tree.data[root])
	// !叶子结点不需要更新lazy
	if root < tree.size {
		tree.lazy[root] = tree.composition(f, tree.lazy[root])
	}
}

func (tree *SegmentTreeRangeAddRangeMinMax) String() string {
	var sb []string
	sb = append(sb, "[")
	for i := 0; i < tree.n; i++ {
		if i != 0 {
			sb = append(sb, ", ")
		}
		sb = append(sb, fmt.Sprintf("%v", tree.Get(i)))
	}
	sb = append(sb, "]")
	return strings.Join(sb, "")
}
